Allow broadcasting in the numerical representations of standard operations (#2609)
PennyLaneAI/pennylane: doc/releases/changelog-dev.md

<h3>New features since last release</h3>

> * Parameter broadcasting within operations and tapes was introduced.
> 

Reviewer: ```suggestion
* Parameter broadcasting within operations and tapes was introduced.
```

Author: :thinking: Not sure what happened here? :)

Reviewer: oh weird, I was trying to delete the line break between the sentence and the PR links :)

  [(#2575)](https://github.com/PennyLaneAI/pennylane/pull/2575)
  [(#2590)](https://github.com/PennyLaneAI/pennylane/pull/2590)
  [(#2609)](https://github.com/PennyLaneAI/pennylane/pull/2609)

>   Parameter broadcasting refers to passing parameters with a (single) leading additional
>   dimension (compared to the expected parameter shape) to `Operator`s.

Reviewer: 💯 
```suggestion
  Parameter broadcasting refers to passing parameters with a (single) leading additional
  dimension (compared to the expected parameter shape) to `Operator`'s.
```

>   Introducing this concept includes multiple changes:

Reviewer: ```suggestion
  Introducing this concept involves multiple changes:
```

  1. New class attributes

>     - `Operator.ndim_params` contains the expected number of dimensions for each parameter
>       of an operator.

Reviewer: ```suggestion
    - `Operator.ndim_params` can be specified by developers to provide the expected number of dimensions for each parameter
      of an operator.
```

>     - `Operator.batch_size` contains the size of an additional parameter-broadcasting axis,
>       if present.
>     - `QuantumTape.batch_size` contains the `batch_size` of its operations (see logic below).

Reviewer: ```suggestion
    - `Operator.batch_size` returns the size of an additional parameter-broadcasting axis,
      if present.
    - `QuantumTape.batch_size` returns the `batch_size` of its operations (see logic below).
```

Reviewer: to try and make clear that these don't need to be defined, whereas `ndim_params` do

  ```pycon
  >>> with qml.tape.QuantumTape() as tape:
  >>>     qml.apply(op)
  >>>     qml.RY(1.9, 0)
  >>> tape.batch_size
  3
  ```
  
  but not `Operation`s with differing (non-`None`) `batch_size`s:

  ```pycon
  >>> with qml.tape.QuantumTape() as tape:
  >>>     qml.apply(op)
  >>>     qml.RY(np.array([1.9, 2.4]), 0)
  ValueError: The batch sizes of the tape operations do not match, they include 3 and 2.
  ```

  When creating a valid broadcasted tape, we can expand it into unbroadcasted tapes with
  the new `broadcast_expand` transform, and execute the three tapes independently.

  ```pycon
  >>> with qml.tape.QuantumTape() as tape:
  >>>     qml.apply(op)
  >>>     qml.RY(1.9, 0)
  >>>     qml.apply(op)
  >>>     qml.expval(qml.PauliZ(0))
  >>> tapes, fn = qml.transforms.broadcast_expand(tape)
  >>> len(tapes)
  3
  >>> dev = qml.device("default.qubit", wires=1)
  >>> fn(qml.execute(tapes, dev, None))
  array([-0.33003414, -0.34999899, -0.38238817])
  ```

  However, devices will handle this automatically under the hood:

  ```pycon
  >>> qml.execute([tape], dev, None)[0]
  array([-0.33003414, -0.34999899, -0.38238817])

>   ```

Reviewer: this changelog entry and example looks amazing @dwierichs!

<h3>New features since last release</h3>

* Many parametrized operations now have the attribute `ndim_params` and
  allow arguments with a broadcasting dimension. Also see entries for #2590
  and #2575 below for details.

>   [(#2609)](https://github.com/PennyLaneAI/pennylane/pull/2609)

Reviewer: Based on this sentence, I would combine all three entries into a singular entry?

  Previously unsupported broadcasted parameters are allowed for example in standard
  rotation gates and matrix operations. The broadcasted dimension is the first dimension
  in numerical operator representations. Note that the broadcasted parameter

>   has to be passed as an `array` but not as a python `list` or `tuple` for most operations.

Reviewer: It might be worth re-phrasing this; the way it is currently worded, it sounds like there is some context missing 🤔 E.g., 'Previously unsupported broadcasted parameters...'

Perhaps the issue are?

- broadcasting in the context of operations and operation parameters has never been defined?
- Starting by saying what _is_ now allowed, rather than what wasn't previously allowed?

  ```pycon
  >>> op = qml.RX(np.array([0.1, 0.2, 0.3], requires_grad=True), 0)
  >>> np.round(op.matrix(), 4)
  tensor([[[0.9988+0.j    , 0.    -0.05j  ],
         [0.    -0.05j  , 0.9988+0.j    ]],
        [[0.995 +0.j    , 0.    -0.0998j],
         [0.    -0.0998j, 0.995 +0.j    ]],
        [[0.9888+0.j    , 0.    -0.1494j],
         [0.    -0.1494j, 0.9888+0.j    ]]], requires_grad=True)
  >>> op.matrix().shape
  (3, 2, 2)

>   ```

Reviewer: This is awesome!
